source/validate.cpp

```diff
@@ -261,6 +273,190 @@ spv_result_t spvValidateIDs(const spv_instruction_t* pInsts,
+  }
+
+  //
+  bool isDefinedId(uint32_t id) const {
+    return defined_ids_.find(id) != std::end(defined_ids_);
+  }
+
+  // Increments the instruction count. Used for diagnostic
+  int incrementInstructionCount() { return instruction_counter_++; }
+
+  libspirv::DiagnosticStream diag(spv_result_t error_code) const {
+    return libspirv::DiagnosticStream({0, 0, (uint64_t)(instruction_counter_)},
+                                      diagnostic_, error_code);
+  }
+
+ private:
+  spv_diagnostic* diagnostic_;
+  // Tracks the number of instructions evaluated by the validator
+  int instruction_counter_;
+
+  // All IDs which have been defined
+  unordered_set<uint32_t> defined_ids_;
+
+  // IDs which have been forward declared but have not been defined
+  unordered_set<uint32_t> unresolved_forward_ids_;
+};
+
+spv_result_t SSAPass(ValidationState_t& _,
+                     function<bool(int)> can_have_forward_declared_ids,
+                     const spv_parsed_instruction_t* inst) {
+  for (int i = 0; i < inst->num_operands; i++) {
+    const spv_parsed_operand_t& operand = inst->operands[i];
+    const spv_operand_type_t& type = operand.type;
+    const uint32_t* operand_ptr = inst->words + operand.offset;
+
+    auto ret = SPV_ERROR_INTERNAL;
+    switch (type) {
+      case SPV_OPERAND_TYPE_RESULT_ID: {
+        _.removeIfForwardDeclared(*operand_ptr);
+        ret = _.definedIds(*operand_ptr);
+        break;
+      }
+      case SPV_OPERAND_TYPE_ID:
+      case SPV_OPERAND_TYPE_TYPE_ID:
+      case SPV_OPERAND_TYPE_MEMORY_SEMANTICS_ID:
+      case SPV_OPERAND_TYPE_SCOPE_ID: {
+        if (_.isDefinedId(*operand_ptr)) {
+          ret = SPV_SUCCESS;
+        } else if (can_have_forward_declared_ids(i)) {
+          ret = _.forwardDeclareId(*operand_ptr);
+        } else {
+          ret = _.diag(SPV_ERROR_INVALID_ID) << "ID " << *operand_ptr
+                                             << " has not been defined";
+        }
+        break;
+      }
+      default:
+        ret = SPV_SUCCESS;
+        break;
+    }
+    if (SPV_SUCCESS != ret) {
+      return ret;
+    }
+  }
+  return SPV_SUCCESS;
+}
+
+template <SpvOp OP>
+bool can_forward_declare(int index) {
+  (void)index;
+  return true;
+}
+
+template <>
+bool can_forward_declare<SpvOpPhi>(int index) {
+  return (1 + index % 2);
+}
+
+template <>
+bool can_forward_declare<SpvOpBranchConditional>(int index) {
+  return index != 0;
+}
+
+template <>
+bool can_forward_declare<SpvOpGroupDecorate>(int index) {
+  return index != 0;
+}
+
+function<bool(int)> getCanBeForwardDeclaredFunction(SpvOp opcode) {
+  function<bool(int index)> out;
+  switch (opcode) {
+    case SpvOpExecutionMode:
+    case SpvOpEntryPoint:
+    case SpvOpFunctionCall:
+    case SpvOpName:
+    case SpvOpMemberName:
+    case SpvOpSelectionMerge:
+
+    // Annotation Instructions
+    case SpvOpDecorate:
+    case SpvOpMemberDecorate:
+    case SpvOpBranch:
+      out = can_forward_declare<SpvOpNop>;
```

Comment:
The use of SpvOpNop is confusing here because OpNop isn't captured by any of the cases that execute this code.  May as well just use the plain lambda inline:   out = [](int){ return true; }